### PR TITLE
Fix IPv6-only bugs on SMTP settings

### LINKED ISSUE #7720: Synapse can't connect to an IPv6-only mail server via hostname
### Description

Synapse only does IPv4 address resolution for the `email.smtp_host` when sending email.
This might be (arguably) an issue on Twisted side though.

Internally, our SMTP server is accessible only with IPv6 (because dual-stack services is not a thing in Kubernetes right now) ; while we could probably work around, I think it's still a bug worth fixing, because IPv6-only environments/clouds are getting more and more traction.

### Steps to reproduce

- have a SMTP host with only an IPv6 (`AAAA`) entry
- configure it as the `email.smtp_host` on Synapse
- try to send an email

Here is the error we got from Sentry:
```
DNSLookupError: DNS lookup failed: postfix.mail.svc.cluster.local.
  File "synapse/handlers/identity.py", line 350, in send_threepid_validation
    await send_email_func(email_address, token, client_secret, session_id)
  File "synapse/push/mailer.py", line 207, in send_add_threepid_mail
    template_vars,
  File "synapse/push/mailer.py", line 323, in send_email
    requireTransportSecurity=self.hs.config.require_transport_security,
```

### Additional informations

I figured that Synapse was only doing `A` resolution by capturing on the container net interface. I wondered why this was the case for SMTP but not in other parts like the federation HTTP client.

It seems that it is because Synapse uses [`twisted.mail.smtp.sendmail`](https://github.com/twisted/twisted/blob/4f49872fe7623eb755d01441a0e5299f6f2363ff/src/twisted/mail/smtp.py#L2093-L2095) to send emails without doing explicit DNS resolution first, and `sendmail` directly does a `reactor.connectTcp` which chooses the address type (IPv4 or IPv6) [based on the host](https://github.com/twisted/twisted/blob/4f49872fe7623eb755d01441a0e5299f6f2363ff/src/twisted/internet/tcp.py#L1538). Because the host supplied to `sendmail` is a hostname and not an IP, Twisted considers it is an IPv4, not an IPv6.

I'm not sure if that should be fixed on Synapse or Twisted side.

### Version information

- **Homeserver**: not matrix.org
- **Version**: 1.15.1
- **Install method**: Kubernetes/docker image
- **Platform**: Containers on bare-metal

cc @ludovicm67, my colleague who found this issue

### PR DESCRIPTION
**Pull request by evilham**
While there, do it in such a fashion that we both document and prepare the groundwork for similar issues relating to direct usage of reactor.connectTCP, which lead to IPv6 incompatibilities.

Closes #7720

### Pull Request Checklist

* [X] Pull request is based on the develop branch
* [X] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [X] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [X] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by evilham** [comment]
This is a more manageable, realistic version of #11508, adding some ground work while also fixing a very specific bug.

File: tests/handlers/test_send_email.py
Comment (clokep) [inline tests/handlers/test_send_email.py:113]:
```suggestion
        super().tearDown()
```

File: tests/handlers/test_send_email.py
Comment (clokep) [inline tests/handlers/test_send_email.py:168]: I think it would make more sense to actually use a [`patch`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) here, we wouldn't need to mess with the teardown at all then.

File: tests/unittest.py
Comment (clokep) [inline tests/unittest.py:322]: Please add documentation into the docstring.

File: tests/unittest.py
Comment (clokep) [inline tests/unittest.py:340]:
```suggestion
        # The homeserver will start connecting places as soon as it starts,
        # so we must update some fake DNS entries early.
```

What connects immediately? Couldn't we do this in the `setUp` of the email tests and simply call `self.reactor.lookups[...] = ...`?

File: tests/unittest.py
Comment (clokep) [inline tests/unittest.py:342]: I suspect these should be `self.lookups`?

```suggestion
        if self.lookups:
            self.reactor.lookups.update(self.lookups)
```

File: synapse/handlers/send_email.py
Comment (clokep) [inline synapse/handlers/send_email.py:122]: I'm failing to find examples of using `TLSMemoryBIOFactory` with `HostnameEndpoint` this way; it seems plausible, but I'm not confident in it.

**Review by clokep — COMMENTED** [review]
Looks somewhat reasonable, but I have a few questions.

File: tests/unittest.py
Comment (evilham) [reply tests/unittest.py:342]: No, this is fine like it is, it is propagating these lookups to the reactor, which uses that dictionary in the `FakeResolver`, see: https://github.com/matrix-org/synapse/blob/84f441f88f51d3f94e1616e1e5507df0dadb6de8/tests/server.py#L468

File: synapse/handlers/send_email.py
Comment (evilham) [reply synapse/handlers/send_email.py:122]: First of all: thanks a lot for the thorough review! Lovely to have issues caught before they land :-).

So, regarding this: you are right that this is something that we see seldom explicitly (we know Twisted's docs can use some help in improving them), but we also use it a lot actually:

`PosixreactorBase.connectSSL` contains [this line](https://github.com/twisted/twisted/blob/e2e113c04ad1ff991f737a7d81a3139512ddc2ab/src/twisted/internet/posixbase.py#L374):

```
    def connectSSL(
        self, host, port, factory, contextFactory, timeout=30, bindAddress=None
    ):
        if tls is not None:
            tlsFactory = tls.TLSMemoryBIOFactory(contextFactory, True, factory)
            return self.connectTCP(host, port, tlsFactory, timeout, bindAddress)
```

If we think about it, that's precisely what we are doing by passing `TLSMemoryBioFactory` this way, given that:

`HostnameEndpoint.connect` contains [this line](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/endpoints.py#L1050):

```
    def connect(self, protocolFactory):
      [...]
                eachAttempt = endpoint.connect(protocolFactory)
```

Where `endpoint` has already been resolved to one of the candidate hosts (directly as IPv{4,6} address).

So, all in all, what we are doing is: instead of using `reactor.connectSSL`, which adds the `TLSMemoryBIOFactory` before actually calling `connectTCP` (which has the host-resolving issues that cause this class of bugs), we use `HostnameEndpoint` and directly use `TLSMemoryBIOFactory` as the wrapper protocol factory.

Hope that makes sense code and theory-wise.

When it comes to actually testing it: it is working :-). I did validate the possible paths with a couple homeservers I control.

File: tests/handlers/test_send_email.py
Comment (evilham) [reply tests/handlers/test_send_email.py:168]: Nice! Good one, it also helps remove the `setUp/tearDown` brainfart, can confirm it works, thank you!

File: tests/unittest.py
Comment (evilham) [reply tests/unittest.py:340]: So, the context for that comment is:

- it was mostly redacted with #11508, which in retrospective was too ambitious and not a right path to contribute to Synapse (incrementally and with actual bug fixing instead of just tests)

So, for example: `BaseMultiWorkerStreamTestCase` forces `self.reactor.lookups` in its `setUp`, as well, and it makes a point that there might be already some Redis connections by the time that part of the code hits.

I'd leave this up to you, we could move the setting of the reactor's lookups to the email test, the logic for doing it here was that other tests could start taking advantage of this mechanism over time.

Will mark as resolved and if you deem it necessary please open it again and we'll deal with it :-).

File: tests/unittest.py
Comment (clokep) [inline tests/unittest.py:318]: Just slightly more readable:

```suggestion
        hijack_auth: Whether to hijack auth to return the user specified
            in user_id.
        lookups: Dictionary of Hostname to IP address (either v4 or v6) that
            will be fed to the reactor before the homeserver is started.
```

File: tests/unittest.py
Comment (clokep) [reply tests/unittest.py:324]: Thinking on this more -- having this be a dictionary will cause mutation across tests, which isn't good for test isolation.

We could probably set it to `None` and initialize to an empty dict in `__init__` if it isn't already set?

Comment (clokep) [reply tests/unittest.py:324]: I suppose the `clear()` in `tearDown` is supposed to handle this, but I think it is cleaner to actually instantiate a new object for each test.

File: tests/server.py
Comment (clokep) [inline tests/server.py:646]: I think `check_ip` isn't doing too much here:

```suggestion
    if isinstance(destination, (address.IPv4Address, address.IPv6Address)):
        try:
            destination.__class__(expected_ip)
        except Exception:
            raise ValueError(
                "Invalid IP type and resolution, got %s, expected: %s %s"
                % (destination, expected_ip, destination.__class__.__name__)
            )
```

**Review by clokep — COMMENTED** [review]
Looks good overall, just a couple of small nits.

File: tests/handlers/test_send_email.py
Comment (clokep) [reply tests/handlers/test_send_email.py:156]: (I don't feel strongly if you think it makes it much clearer, but I find it one additional level of indirection needed to track what's going on.)

File: tests/server.py
Comment (evilham) [reply tests/server.py:646]: I think I got the gist of what you are hinting at, let me know if the changes I'm about to commit address your concerns.

However, do note the comment (and let me know if it can be made clearer):
```
    # We use address.IPv{4,6}Address to check what the reactor thinks it is
    # is sending but check for validity with with ipaddress.IPv{4,6}Address
    # because they fail with IPs on the wrong network stack.
```

Twisted's `address.IPv{4,6}Address` classes do not actually perform any kind of check, hence why the check goes explicitly against `isinstance(destination, address.IPv4Address` but the validation goes against `ipaddress.IPv4Address`.

Here is a quick interactive shell sample of what I mean:
```
>>> from twisted.internet import address
>>> address.IPv4Address("TCP", host="::1", port=1234)
IPv4Address(type='TCP', host='::1', port=1234)
>>> address.IPv6Address("TCP", host="127.0.0.1", port=1234)
IPv6Address(type='TCP', host='127.0.0.1', port=1234, flowInfo=0, scopeID=0)
>>>
>>> import ipaddress
>>> ipaddress.IPv4Address("::1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/ipaddress.py", line 1304, in __init__
    self._ip = self._ip_int_from_string(addr_str)
  File "/usr/local/lib/python3.9/ipaddress.py", line 1191, in _ip_int_from_string
    raise AddressValueError("Expected 4 octets in %r" % ip_str)
ipaddress.AddressValueError: Expected 4 octets in '::1'
>>> ipaddress.IPv6Address("127.0.0.1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/ipaddress.py", line 1915, in __init__
    self._ip = self._ip_int_from_string(addr_str)
  File "/usr/local/lib/python3.9/ipaddress.py", line 1628, in _ip_int_from_string
    raise AddressValueError(msg)
ipaddress.AddressValueError: At least 3 parts expected in '127.0.0.1'
```

Note that the errors from ipaddress are super terrible, and that's why we raise a different exception.

By forcing the exception to happen in my validations I see following instead:

```
  File "tests/server.py", line 645, in validate_connector
    raise ValueError(
builtins.ValueError: Invalid IP type and resolution for IPv4Address(type='TCP', host='localhost', port=465). Expected ::1 to be IPv4Address
```

I think that would be at least somewhat helpful if a developer hits this failure in the test suite, particularly since that would bring them to `validate_connector`, whose comment along with this error message would bring them along the right path.

Comment (clokep) [reply tests/server.py]: The difference between `address.IPv4Address` and `ipaddress.IPv4Address` was completely lost to me -- this needs to be made much more obvious.

Comment (evilham) [reply tests/server.py]: This still holds with the modified function? (GH is showing the old code here)

```
    # We use address.IPv{4,6}Address to check what the reactor thinks it is
    # is sending but check for validity with ipaddress.IPv{4,6}Address
    # because they fail with IPs on the wrong network stack.
    cls_mapping = {
        address.IPv4Address: ipaddress.IPv4Address,
        address.IPv6Address: ipaddress.IPv6Address,
    }

    cls = cls_mapping.get(destination.__class__)
```

File: synapse/handlers/send_email.py
Comment (clokep) [inline synapse/handlers/send_email.py:121]: Last comment, sorry. 😄 

I think it would make sense to move the `endpoint = HostnameEndpoint(...)` below the `if force_tls` bit, as then all the code to figure out which "factory" we want to use is next to each other.

**Review by clokep — APPROVED** [review]
Looks good, minus one last comment!

File: synapse/handlers/send_email.py
Comment (evilham) [reply synapse/handlers/send_email.py:121]: Heh, that's a great point!

**Comment by clokep** [comment]
Thanks for handling this! 🎉

**Comment by glyph** [comment]
yaaay, super happy to see more `HostnameEndpoint` adoption!

**Pull request merged**
